### PR TITLE
Efficient reading

### PR DESCRIPTION
**Pull request by clusterfudge**
Avoid using streaming read where possible.

### CODE REVIEW

File: webapp/graphite/readers/remote.py [hunk 1]
```diff
@@ -106,3 +111,65 @@ def fetch_multi(self, startTime, endTime, now=None, requestContext=None):
...
+    # Here be Snap monkey-patch code
+    #
+    def deserialize(self, result):
+        # avoid buffered reader if possible, instead using in-memory unpacking for small->REMOTE_BUFFER_SIZE payloads
+        try:
+            start = time.time()
+            measured_reader = MeasuredReader(BufferedHTTPReader(result, settings.REMOTE_BUFFER_SIZE))
+            first_chunk = self._fill_first_buffer(measured_reader, settings.REMOTE_BUFFER_SIZE)
+            if len(first_chunk) < settings.REMOTE_BUFFER_SIZE:
+                log.info("Using inline deserializer for small payload")
```
Comment (deniszh) [inline webapp/graphite/readers/remote.py]: I think it should be `log.debug` - here, and in other places too.

**Comment by deniszh** [comment]
@clusterfudge @DanCech @iksaif - what we going to do with this change?
Reminding that is related to the problem described in https://github.com/graphite-project/graphite-web/issues/2313#issuecomment-401876110
I also notice some performance degradation on big requests but I can't prove it with numbers, though.
OTOH I don't really like consequences of mandatory REMOTE_BUFFER_SIZE. Can we probably make avoiding streaming deserialization optional somehow? I.e if REMOTE_BUFFER_SIZE > 0 use it, if not - fallback to stream implementation?

**Comment by iksaif** [comment]
I agree with making REMOTE_BUFFER_SIZE optional

**Comment by clusterfudge** [comment]
@deniszh yeah, I like the idea of 0 to disable (vs using an absurdly high number to avoid). I'll refresh this PR to reflect.

**Comment by deniszh** [comment]
Hi @clusterfudge ,
Could you please fix this PR soon, if you want that change still be included in upcoming 1.1.4 release. Thanks!

**Comment by clusterfudge** [comment]
Hey @deniszh , I spent a bit of time polishing yesterday and should be able to get it out tonight. What's the target cutoff for 1.1.4?

**Comment by deniszh** [comment]
Soon ™️ :)

**Comment by deniszh** [comment]
@clusterfudge : OK, you mentioned before that if request size is more then REMOTE_BUFFER then it's almost guaranteed to fail, but I'm not really getting why. Is it still true?
If yes - then we need to inform users about changing behavior.

**Comment by clusterfudge** [comment]
@deniszh That was an artifact of the prior implementation and is no longer relevant. In this implementation, we use the streaming deserializers if we have REMOTE_BUFFER_SIZE > 0, otherwise we use the inline deserialization. This does not change the behavior of existing users, but allows people to opt-out of streaming serialization to increase performance on large payloads.
TL/DR; that comment is no longer relevant.

**Comment by deniszh** [comment]
Ah, that's what I suspect - but thanks for the clarification.
Merging.

**Comment by Quentin-M** [comment]
This PR appears to break metrics discovery: https://github.com/graphite-project/graphite-web/issues/2352.

**Pull request merged**